extensions/amp-story/0.1/amp-story-page.js

```diff
@@ -0,0 +1,642 @@
+ * @const {string}
+ */
+const PAGE_SHOWN_CLASS_NAME = 'i-amphtml-story-page-shown';
+
+
+/**
+ * The duration of time (in milliseconds) to show the loading screen for this
+ * page, before showing the page content.
+ * @const {number}
+ */
+const LOAD_TIMEOUT_MS = 8000;
+
+
+/**
+ * The delay (in milliseconds) to wait between polling for loaded resources.
+ */
+const LOAD_TIMER_POLL_DELAY_MS = 250;
+
+
+
+/** @private @enum {!RegExp} */
+const TIME_REGEX = {
+  MILLISECONDS: /^(\d+)ms$/,
+  SECONDS: /^(\d+)s$/,
+};
+
+
+/** @private @const {string} */
+const TAG = 'amp-story-page';
+
+
+
+export class AmpStoryPage extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private @const {?AnimationManager} */
+    this.animationManager_ = null;
+
+    /** @private @const {!Array<!PageElement>} */
+    this.pageElements_ = [];
+
+    /** @private {?function()} */
+    this.resolveLoadPromise_ = null;
+
+    /** @private {?Promise<undefined>} */
+    this.loadPromise_ = new Promise(resolve => {
+      this.resolveLoadPromise_ = resolve;
+    });
+
+    /** @private {?Promise<undefined>} */
+    this.loadTimeoutPromise_ = null;
+
+    /** @private @const {!../../../src/service/timer-impl.Timer} */
+    this.timer_ = Services.timerFor(this.win);
+
+    /** @private {boolean} */
+    this.isLoaded_ = false;
+
+    /** @private {?UnlistenDef} */
+    this.autoAdvanceUnlistenDef_ = null;
+  }
+
+
+  /*
+   * @return {?./animation.AnimationManager}
+   */
+  maybeCreateAnimationManager_() {
+    if (!this.animationManager_) {
+      if (!hasAnimations(this.element)) {
+        return;
+      }
+
+      this.animationManager_ = AnimationManager.create(
+          this.element, this.getAmpDoc(), this.getAmpDoc().getUrl());
+    }
+  }
+
+
+  /** @override */
+  buildCallback() {
+    upgradeBackgroundAudio(this.element);
+    this.markMediaElementsWithPreload_();
+    this.maybeCreateAnimationManager_();
+    this.initializeLoading_();
+  }
+
+
+  /**
+   * Marks any AMP elements that represent media elements with preload="auto".
+   */
+  markMediaElementsWithPreload_() {
+    const mediaSet = this.element.querySelectorAll('amp-audio, amp-video');
+    for (const mediaItem of mediaSet) {
+      mediaItem.setAttribute('preload', 'auto');
+    }
+  }
+
+
+  /**
+   * Initializes the loading screen for this amp-story-page, and the listeners
+   * to remove it once loaded.
+   */
+  initializeLoading_() {
+    // Add the loading screen into the DOM.
+    const loadingScreen = this.win.document.createElement('div');
+    loadingScreen.classList.add('i-amphtml-story-page-loading-screen');
+    loadingScreen./*OK*/innerHTML = LOADING_SCREEN_CONTENTS_TEMPLATE;
+    this.element.appendChild(loadingScreen);
+
+    // Build a list of page elements and poll until they are all loaded.
+    this.pageElements_ = PageElement.getElementsFromPage(this);
```

Comment:
Do we know how performant this is in practice?

Reply:
There's a timeout (8s), but in practice it seems to be working reasonably, depending on the elements on the page.

We need to do something like this, though; having all the descendants loaded asynchronously whenever they are ready doesn't work well for this format.  A potential future optimization is to use the `hero` attribute (or similar) to allow publishers to mark up which element(s) to wait for, but I think we'll need to default to a conservative approach, as partially loaded pages are very problematic for the format.

Reply:
8ms I hope? 😅  Is that on a workstation with CPU throttling or on a low-end mobile device?

Reply:
Sorry, I guess it depends which part you're asking about the performance of.  I don't think we've evaluated the performance of the `PageElement.getElementsFromPage(this)` call, but if you're talking about the comment `// Build a list of page elements and poll until they are all loaded.`, then that's what has the 8s timeout.  We'll show a loading indicator/screen for up to 8s after the user has started looking at the screen.  It takes so long because we're actually waiting for the assets (i.e. mostly videos) to be fetched over the network.

Reply:
I see. I was actually curious about the DOM queries. Generally we try to avoid DOM scans in favor of custom element API. For instance, forms would be more robust to dynamic markup changes  (e.g. new form elements created from amp-list) if it used an `<amp-form>` instead of `<form>` as it does now.

We could devise a kind of dynamic add-on system to avoid the need for scanning here. Only justifiable if the scan is slow or we foresee dynamic markup in STAMP.